### PR TITLE
[ Flatten ] backwarding @open sesame 06/18 14:15

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR includes back propagation of Flatten Layer.
. backwarding Flatten Layer.
. batch(), channel(), width(), height(), setDim() of tensor
. unit test of flatten layer

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/include/flatten_layer.h [hunk 1]
```diff
@@ -86,6 +87,14 @@ class FlattenLayer : public Layer {
    */
   Tensor backwarding(Tensor in, int iteration);
 
+  /**
+   * @brief     set Property of layer
+   * @param[in] values values of property
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  int setProperty(std::vector<std::string> values) { return ML_ERROR_NONE; };
```
Comment (kparichay) [inline nntrainer/include/flatten_layer.h:96]: How about returning `ML_ERROR_NOT_SUPPORTED` leading to an error when trying to set property, or giving out a warning that this property is being ignored?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 2]
```diff
@@ -601,6 +602,79 @@ TEST(nntrainer_Pooling2D, forwarding_01_p) {
...
+  std::vector<std::string> input_str;
+  nntrainer::TensorDim previous_dim;
+  previous_dim.setTensorDim("1:2:4:4");
+  layer.setInputDimension(previous_dim);
+
+  status = layer.initialize(false);
+  EXPECT_EQ(status, ML_ERROR_NONE);
+
+  nntrainer::Tensor in(1, 1, 1, 32);
+  nntrainer::Tensor out, result(1, 1, 1, 32);
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:656]: how about also creating a unit test with batch size more than 1 to test that flatten does not affect the batch size?

Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_layers.cpp:656]: You are right it does not affect the batch. However the more unit test is the better. So I will add more test cases. ( and negative test cases as well )

File: nntrainer/include/flatten_layer.h (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/include/flatten_layer.h]: fixed. Thanks.

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 2 above)
Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_layers.cpp:783]: some more unit tests are added except negative test cases.

File: debian/rules [hunk 3]
```diff
@@ -35,7 +35,7 @@ override_dh_auto_test:
 	tar xzf ./packaging/trainset.tar.gz -C build
 	tar xzf ./packaging/valset.tar.gz -C build
 	tar xzf ./packaging/testset.tar.gz -C build
-	tar xzf ./packaging/conv2d_unittest.tar.gz -C build
+	tar xzf ./packaging/unittest_layers.tar.gz -C build
```
Comment (zhoonit) [inline debian/rules:38]: Could you also update `../unittest/meson.build:4`  `unzip_target` please?

Comment (jijoongmoon) [reply debian/rules:38]: right. fixed. thanks.
